src/Core/Content/Category/Service/DefaultCategoryLevelLoader.php

```diff
@@ -0,0 +1,225 @@
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Contracts\Cache\ItemInterface;
+use Symfony\Contracts\Cache\TagAwareCacheInterface;
+
+/**
+ * @internal
+ */
+#[Package('discovery')]
+class DefaultCategoryLevelLoader implements EventSubscriberInterface
+{
+    private const CACHE_TAG = 'category_level_loader';
+
+    /**
+     * @param SalesChannelRepository<CategoryCollection> $categoryRepository
+     */
+    public function __construct(
+        private readonly TagAwareCacheInterface $cache,
+        private readonly EventDispatcherInterface $eventDispatcher,
+        private readonly SalesChannelRepository $categoryRepository,
+    ) {
+    }
+
+    /**
+     * @return array<string, string>
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CategoryEvents::CATEGORY_WRITTEN_EVENT => 'invalidateCache',
+        ];
+    }
+
+    public function loadLevels(
+        string $rootId,
+        int $rootLevel,
+        SalesChannelContext $context,
+        Criteria $criteria,
+        int $depth,
+    ): CategoryCollection {
+        if ($context->getSalesChannel()->getNavigationCategoryId() === $rootId) {
+            return $this->cached(
+                $rootId,
+                $rootLevel,
+                $context,
+                $criteria,
+                $depth,
+            );
+        }
+
+        return $this->load(
+            $rootId,
+            $rootLevel,
+            $context,
+            $criteria,
+            $depth,
+        );
+    }
+
+    public function invalidateCache(): void
+    {
+        $this->cache->invalidateTags([self::CACHE_TAG]);
+    }
+
+    private function cached(
+        string $rootId,
+        int $rootLevel,
+        SalesChannelContext $context,
+        Criteria $criteria,
+        int $depth,
+    ): CategoryCollection {
+        $cacheKey = $this->getCacheKey($rootId, $context, $criteria, $depth);
+
+        if ($cacheKey === null) {
+            return $this->load($rootId, $rootLevel, $context, $criteria, $depth);
+        }
+
+        $cacheMetadata = [
+            ItemInterface::METADATA_TAGS => self::CACHE_TAG,
+        ];
+
+        $compressed = $this->cache->get(
+            $cacheKey,
+            fn () => CacheValueCompressor::compress($this->load($rootId, $rootLevel, $context, $criteria, $depth)),
+            null,
+            $cacheMetadata
+        );
+
+        $categories = CacheValueCompressor::uncompress($compressed);
+        \assert($categories instanceof CategoryCollection);
+
+        return $categories;
+    }
+
+    private function getCacheKey(string $rootId, SalesChannelContext $context, Criteria $criteria, int $depth): ?string
+    {
+        $event = new CategoryLevelLoaderCacheKeyEvent(
+            [
```

Comment:
why do you put them into an array? ah I see, those are the values, which are used for the hash. :thinking:

Reply:
has those are the parts that will form the hash, plugins can add parts and we can also change them, you should not rely on the content of the array, that's why the important information is also available directly in the event